### PR TITLE
Updated Models/Encoding Explorations

### PR DESCRIPTION
**Pull request by oliver-adams-b**
Created a quick 'pythonic' [file transformer](https://github.com/oliver-adams-b/library/blob/main/egyptian_fruit_bat/data_wranglers/dataset_transformer.py) so that I could speed up preprocessing. This class takes a function (with a fixed/predetermined domain and range), and applies this function onto a provided dataset, and saves the result into a dataset with the same file tree as the input. I used this class just to transform the EFB audio dataset into a dataset of images. 

Made a custom model with some [positional encoding](https://github.com/oliver-adams-b/library/blob/main/egyptian_fruit_bat/efb_context_labeler_pos_enc_comparison.ipynb) (or my implementation of positional encoding). Using a custom layer, I train models with and without positional encoding and plot the distribution of the inner layer activations (positional encoding _might_ perform slightly better!)

Created a [VAE](https://github.com/oliver-adams-b/library/blob/main/egyptian_fruit_bat/efb_vae.ipynb) using the positional encoding layer stuff -- looking good so far, just need to train it a little more I think. Very exciting to see the progress of the model during training!

### CODE REVIEW

**Comment by radekosmulski** [comment]
Hey Oliver!

Thank you for the PR. Really excited to see that you are continuing to explore these methods 🙂

A couple of comments:
* for the model with positional encoding, seems the comparison is done with different learning rates between the models:
image
(1e-3 vs 1e-2) - did you observe 1e-2 perform better for the model without positional encoding than 1e-3
* the umap plots are very helpful 🙂
* nice attempt at VAEs! looking at the results, doesn't seem it is learning much though? or am I reading this wrong?
* interesting exploration of the rainbowification 👍
* are all the models trained on the same subset of data? meaning, can I compare the results between rainbowification and positional encodings?
* nice, thorough exploration of the differences between labels

I do appreciate your notebooks and your results. Internally, we are having a discussion about what direction for the library we should take. Some of the notebooks that you share do not demonstrate fundamental techniques, rather, you are going beyond that into the realm of exploration. This is great for learning and genuinely interesting, but I am not sure if this is what would work well as the content for the library, or whether a better idea would be to link to your work from the library where it would live under your github account. Given that the methods you explore are more experimental in nature, I would probably lean towards the latter.

**Comment by radekosmulski** [comment]
Sorry, only now noticed the subfolder `yossi_classification_tasks` and also the new notebook on unet completion. Seems given the variability of the results from the models on classification tasks it was a good idea to portray the distribution of results 👍, love the graphs BTW. Significant improvement on gender classification vs the paper.

The unet for completion seems like a nice exploration, but I am not sure there are any findings that we could extract from it or to what extent this can be useful for bioacoustic research, at least in the current form? Nice exploration nonetheless!
